nano/node/distributed_work.cpp

```diff
@@ -0,0 +1,372 @@
+			{
+				node.unresponsive_work_peers = false;
+				set_once (work);
+				stop (true);
+			}
+			else
+			{
+				node.logger.try_log (boost::str (boost::format ("Incorrect work response from %1% for root %2% with diffuculty %3%: %4%") % address % root.to_string () % nano::to_string_hex (difficulty) % work_text));
+				handle_failure (last);
+			}
+		}
+		else
+		{
+			node.logger.try_log (boost::str (boost::format ("Work response from %1% wasn't a number: %2%") % address % work_text));
+			handle_failure (last);
+		}
+	}
+	catch (...)
+	{
+		node.logger.try_log (boost::str (boost::format ("Work response from %1% wasn't parsable: %2%") % address % body_a));
+		handle_failure (last);
+	}
+}
+
+void nano::distributed_work::set_once (boost::optional<uint64_t> work_a)
+{
+	if (!completed.exchange (true))
+	{
+		callback (work_a);
+	}
+}
+
+void nano::distributed_work::failure (boost::asio::ip::address const & address)
+{
+	auto last (remove (address));
+	handle_failure (last);
+}
+
+void nano::distributed_work::handle_failure (bool const last)
+{
+	if (last && !completed)
+	{
+		node.unresponsive_work_peers = true;
+		if (!local_generation_started)
+		{
+			if (stopped)
+			{
+				callback (boost::none);
+			}
+			else
+			{
+				if (backoff == 1 && node.config.logging.work_generation_time ())
+				{
+					node.logger.always_log ("Work peer(s) failed to generate work for root ", root.to_string (), ", retrying...");
+				}
+				auto now (std::chrono::steady_clock::now ());
+				auto root_l (root);
+				auto callback_l (callback);
+				std::weak_ptr<nano::node> node_w (node.shared ());
+				auto next_backoff (std::min (backoff * 2, (unsigned int)60 * 5));
+				// clang-format off
+				node.alarm.add (now + std::chrono::seconds (backoff), [ node_w, root_l, callback_l, next_backoff, difficulty = difficulty ] {
+					if (auto node_l = node_w.lock ())
+					{
+						node_l->distributed_work.make (next_backoff, root_l, callback_l, difficulty);
+					}
+				});
+				// clang-format on
+			}
+		}
+	}
+}
+
+bool nano::distributed_work::remove (boost::asio::ip::address const & address)
+{
+	std::lock_guard<std::mutex> guard (mutex);
+	outstanding.erase (address);
+	return outstanding.empty ();
+}
+
+nano::distributed_work_factory::distributed_work_factory (nano::node & node_a) :
+node (node_a)
+{
+}
+
+void nano::distributed_work_factory::make (nano::block_hash const & root_a, std::function<void(boost::optional<uint64_t>)> const & callback_a, uint64_t difficulty_a)
+{
+	make (1, root_a, callback_a, difficulty_a);
+}
+
+void nano::distributed_work_factory::make (unsigned int backoff_a, nano::block_hash const & root_a, std::function<void(boost::optional<uint64_t>)> const & callback_a, uint64_t difficulty_a)
+{
+	auto distributed (std::make_shared<nano::distributed_work> (backoff_a, node, root_a, callback_a, difficulty_a));
+	{
+		std::lock_guard<std::mutex> guard (mutex);
+		work[root_a].emplace_back (distributed);
+	}
+	distributed->start ();
+	cleanup ();
```

Comment:
Would it make sense to call `cleanup` first, so you don't waste time iterating over the newly created distributed work? (although it's probably negligible) Also we generally use the term `cleanup ()` to indicate it's basically annihilating the whole object. Perhaps add a suffix or use a different term to indicate it's only a partial cleanup, `cleanup_expired`, `purge_finished_work` or something.

Reply:
Makes sense! Changing to `cleanup_finished` 👍